Add capability to render to cpu buffer or double buffer

DGriffin91 opened the pull request:
# Objective

Provides a portion of the functionally needed for https://github.com/bevyengine/bevy/issues/22

The goal of the PR is to provide a straightforward way to:
- Render to a cpu buffer that can be used to output images files.
- Render to a double buffer, allowing the image to be easily used as a texture in the same render layer.
- Simplify rendering to a texture. Doing so without having to directly manipulate the render graph.

There are three examples showing each of these respective methods:
`examples/3d/frame_capture_cpu.rs`
`examples/3d/frame_capture_double_buffer.rs`
`examples/3d/frame_capture_gpu.rs`

The `FrameCapturePlugin` is also setup to allow use of multiple, simultaneous, renders using a variety of targets.

The location of the `FrameCapturePlugin` was lightly discussed on discord. There isn't currently a great place for utility plugins like this one.

alice-i-cecile commented on examples/3d/frame_capture_cpu.rs:
```diff
@@ -0,0 +1,117 @@
+use bevy::render::renderer::RenderDevice;
+
+#[derive(Component, Default)]
+pub struct CaptureCamera1;
```
Is there a reason why this needs to be numbered?

DGriffin91 replied:
Just a left over from when I was testing using multiple cameras. I'll ditch the number.

alice-i-cecile commented:
Super cool! A few comments on the examples:

1. Please add quick module level doc comments at the top of each explaining what their trying to do.
2. It feels very awkward that these are in the 3D folder. There's fundamentally nothing in them that wouldn't work for 2D. @mockersf, are you on board with creating a `rendering` examples folder? I know we've discussed this before, but this is a clear case of the need.
3. Can we save the rendered textures to file? If so, that would be incredibly valuable to demonstrate in an example in [the root level tests folder](https://github.com/bevyengine/bevy/tree/main/tests)

DGriffin91 commented:
@alice-i-cecile 
> 1. Please add quick module level doc comments at the top of each explaining what their trying to do.

Just added some doc comments. Is what I added the sort of thing you are looking for?
> 2. It feels very awkward that these are in the 3D folder.

Agreed. I'll have to do some testing to make sure these can work in 2D as well.
> 3. Can we save the rendered textures to file?

Yep! That's what the `frame_capture_cpu` example currently does, and is part of the motivation for adding this feature.

alice-i-cecile commented:
WRT the comments, see https://github.com/bevyengine/bevy/pull/4438 :)

mockersf commented on crates/bevy_camera/Cargo.toml:
```diff
@@ -0,0 +1,25 @@
+
+[dependencies]
+# bevy
+bevy_app = { path = "../bevy_app", version = "0.8.0-dev" }
+bevy_asset = { path = "../bevy_asset", version = "0.8.0-dev" }
+bevy_ecs = { path = "../bevy_ecs", version = "0.8.0-dev" }
+bevy_utils = { path = "../bevy_utils", version = "0.8.0-dev" }
+bevy_render = { path = "../bevy_render", version = "0.8.0-dev" }
+bevy_math = { path = "../bevy_math", version = "0.8.0-dev" }
+bevy_core = { path = "../bevy_core", version = "0.8.0-dev" }
+bevy_core_pipeline = { path = "../bevy_core_pipeline", version = "0.8.0-dev" }
```
could you reorder them alphabetically and check that they are all needed? I'm not sure for bevy_utils, bevy_math or bevy_core

mockersf commented on crates/bevy_camera/src/frame_capture.rs:
```diff
@@ -0,0 +1,363 @@
+
+        let gpu_image = images.add(image);
+
+        let padded_bytes_per_row = RenderDevice::align_copy_bytes_per_row(width as usize) * 4;
```
I'm not happy about the `* 4` there (and the other time there's a 4). Some texture formats have less than 4 components, and some uses more than one byte per component.
Should this be `* format.describe().components as usize * format.describe().block_size as usize` instead?

DGriffin91 replied:
Good catch.

Based on this: [ImageDataLayout](https://wgpu.rs/doc/wgpu/struct.ImageDataLayout.html) I think it may be:

```rs
let padded_bytes_per_row = RenderDevice::align_copy_bytes_per_row(
    (gpu_image.size.width as usize / format.block_dimensions.0 as usize) * format.block_size as usize,
);
```

[This](https://github.com/bevyengine/bevy/blob/00f83941b16e7a175c5c1503be81a648e23d81df/crates/bevy_render/src/texture/image.rs#L582) is what Bevy does for `prepare_asset` in `image.rs`

And it looks like `pixel_size()` just does `info.type_size * info.num_components`. I'm not sure if this ends up being always correct if the texture is compressed.

There's also the issue of the buffer being larger than the user might expect since it's [aligned to 256 bytes](https://docs.rs/wgpu-types/latest/wgpu_types/constant.COPY_BYTES_PER_ROW_ALIGNMENT.html)

This is fine for both GPU examples since it's what the gpu expects on both ends. But the one that copies to the CPU results in row padded data.

Do we crop the buffer for the user? If so, I'm not sure exactly what the most efficient way of doing that is. Each row is potentially too long, so we can't just slice off the end of the flat buffer.

mockersf commented on crates/bevy_camera/src/frame_capture.rs:
```diff
@@ -0,0 +1,363 @@
+            commands
+                .get_or_spawn(entity)
+                .insert_bundle((
+                    RenderPhase::<Opaque3d>::default(),
+                    RenderPhase::<AlphaMask3d>::default(),
+                    RenderPhase::<Transparent3d>::default(),
```
would it be possible to have it work also for 2d / ui?

mockersf commented on examples/3d/frame_capture_cpu.rs:
```diff
@@ -0,0 +1,119 @@
+                bevy::camera::TargetBuffer::CPUBuffer(target_buffer) => {
+                    target_buffer.get(&render_device, |buf| {
+                        image::save_buffer(
+                            format!("../test{i}.png"),
```
this will write every frame to the same file, with a good chance of leaving it corrupted (with 120 writes per second on my laptop)

could this system be put on a tilmestep, maybe every second, and changing the filename every time?

alice-i-cecile commented:
@DGriffin91 does this fix #22? If so, please add it to the PR description.

DGriffin91 commented:
@alice-i-cecile It should at least help with the "copy texture data from the gpu to an image file" portion. I just tried making a headless example and haven't been able to.

If I manually add plugins:
```
.add_plugin(CorePlugin)
.add_plugin(TransformPlugin)
.add_plugin(HierarchyPlugin)
.add_plugin(AssetPlugin)
.add_plugin(ScenePlugin)
.add_plugin(RenderPlugin)
.add_plugin(CorePipelinePlugin)
.add_plugin(GltfPlugin)
.add_plugin(PbrPlugin)
```
I get an error on this line:
https://github.com/bevyengine/bevy/blob/328c26d02c50de0bc77f0d24a376f43ba89517b1/crates/bevy_render/src/lib.rs#L128
```rust
thread 'main' panicked at 'Requested resource bevy_window::windows::Windows does not exist in the `World`. 
                Did you forget to add it using `app.add_resource` / `app.init_resource`?
                Resources are also implicitly added via `app.add_event`,
                and can be added by plugins.', crates\bevy_render\src\lib.rs:128:41
```
And if I try the wgpu `backend: None` from the `headless_defaults.rs` with:
```rust
.insert_resource(WgpuSettings {
    backends: None,
    ..default()
})
.add_plugins(DefaultPlugins)
```
I get the error: 
```
thread 'main' panicked at 'Sub-App with label 'RenderApp' does not exist', 
[...]\bevy_app\src\app.rs:884:27
```

So I'm not sure if headless rendering in generally possible in bevy yet. And I also don't know if making headless rendering possible is in the scope of this PR.

alice-i-cecile commented:
Sounds good. Headless rendering has been a bit of a headache for a long time, so that doesn't surprise me. Fixing that is definitely out of scope.

mockersf commented:
waiting on the "high level camera driven render targets" PR

YoshieraHuang commented:
> @alice-i-cecile It should at least help with the "copy texture data from the gpu to an image file" portion. I just tried making a headless example and haven't been able to.
> 
> If I manually add plugins:
> 
> ```rust
> .add_plugin(CorePlugin)
> .add_plugin(TransformPlugin)
> .add_plugin(HierarchyPlugin)
> .add_plugin(AssetPlugin)
> .add_plugin(ScenePlugin)
> .add_plugin(RenderPlugin)
> .add_plugin(CorePipelinePlugin)
> .add_plugin(GltfPlugin)
> .add_plugin(PbrPlugin)
> ```
> 
> I get an error on this line:
> 
> https://github.com/bevyengine/bevy/blob/328c26d02c50de0bc77f0d24a376f43ba89517b1/crates/bevy_render/src/lib.rs#L128
> 
> ```rust
> thread 'main' panicked at 'Requested resource bevy_window::windows::Windows does not exist in the `World`. 
>                 Did you forget to add it using `app.add_resource` / `app.init_resource`?
>                 Resources are also implicitly added via `app.add_event`,
>                 and can be added by plugins.', crates\bevy_render\src\lib.rs:128:41
> ```
> 
> And if I try the wgpu `backend: None` from the `headless_defaults.rs` example:
> 
> ```rust
> .insert_resource(WgpuSettings {
>     backends: None,
>     ..default()
> })
> .add_plugins(DefaultPlugins)
> ```
> 
> I get the error:
> 
> ```
> thread 'main' panicked at 'Sub-App with label 'RenderApp' does not exist', 
> [...]\bevy_app\src\app.rs:884:27
> ```
> 
> So I'm not sure if headless rendering in generally possible in bevy yet. And I also don't know if making headless rendering possible is in the scope of this PR.

Only the `WinitPlugin` should be disable to prevent the windowing and associated rendering. This scripts works with `frame_capture_cpu` example:
```
App::new()
...
        .add_plugins_with(DefaultPlugins, |group| group.disable::<WinitPlugin>() )
...
```

However, only one frame is rendered and then the app exits, since `WinitPlugin` provides built-in eventloop to trigger app to run forever.

mockersf commented:
@DGriffin91 https://github.com/bevyengine/bevy/pull/4898 has been merged, could you update this PR? I would love to see it move forward.

zimond commented:
I tested this PR on 2d camera and it just works. But I think the whole `extract_camera_phases` thing is creating redundant code, as it is identical to what is done for 2d/3d camera

DGriffin91 commented:
> I tested this PR on 2d camera and it just works (by changing 3d subgraph rendering to 2d, etc). I think maybe the whole `extract_camera_phases` thing is creating redundant code, as it is identical to what is done for 2d/3d camera

If I remove this (below) from `extract_camera_phases` with the `frame_capture_cpu` example, it doesn't output the 3D model and I just get a blank grey png.
```rust
.insert_bundle((
    RenderPhase::<Opaque3d>::default(),
    RenderPhase::<AlphaMask3d>::default(),
    RenderPhase::<Transparent3d>::default(),
))
```

zimond commented:
@DGriffin91 Yes, I didn't mean it's unnecessary, it's just a duplication of  what is done in [camera3d pipeline](https://github.com/bevyengine/bevy/blob/5e2cfb2f19e0e2287fd48d55df92f78617acf831/crates/bevy_core_pipeline/src/core_3d/mod.rs#L187)

DGriffin91 commented:
@zimond Gotcha, that makes sense. And thank you! We are discussing how to get this setup with the new camera driven stuff on discord here in case you are interested: https://discord.com/channels/691052431525675048/986788850535915540/

superdump commented on crates/bevy_core_pipeline/src/core_3d/main_pass_3d_node.rs:
```diff
@@ -194,6 +199,31 @@ impl Node for MainPass3dNode {
             }
         }
 
+        // TODO: should this live here or in a separate node?
```
Separate node imo.

superdump commented on crates/bevy_render/src/camera/camera.rs:
```diff
@@ -247,6 +247,8 @@ pub enum RenderTarget {
     Window(WindowId),
     /// Image to which the camera's view is rendered.
     Image(Handle<Image>),
+    /// Buffered Image to which the camera's view is rendered.
```
I feel like this description could be clearer. The first image is used for rendering into and is then copied into the second image, right?

superdump commented on examples/3d/render_to_double_buffer.rs:
```diff
@@ -0,0 +1,166 @@
+            format: TextureFormat::Rgba8UnormSrgb,
+            mip_level_count: 1,
+            sample_count: 1,
+            usage: TextureUsages::TEXTURE_BINDING
+                | TextureUsages::COPY_DST
+                | TextureUsages::COPY_SRC
+                | TextureUsages::RENDER_ATTACHMENT,
```
I think this one only needs RENDER_ATTACHMENT | COPY_SRC.

superdump commented on examples/3d/render_to_double_buffer.rs:
```diff
@@ -0,0 +1,166 @@
+            format: TextureFormat::Rgba8UnormSrgb,
+            mip_level_count: 1,
+            sample_count: 1,
+            usage: TextureUsages::TEXTURE_BINDING
+                | TextureUsages::COPY_DST
+                | TextureUsages::COPY_SRC
+                | TextureUsages::RENDER_ATTACHMENT,
```
I think this one only needs COPY_DST | TEXTURE_BINDING.

paulkre commented:
I'm looking forward to using this feature. I hope it gets merged soon!

paulkre commented:
Is it possible to save a PNG image with the `wgpu::Texture` struct? I only found a way to do this with the `wgpu::Buffer` which requires the use of `wgpu::CommandEncoder::copy_texture_to_buffer` method.

DGriffin91 commented:
@paulkre Check out this newer PR. It outputs PNG images: https://github.com/bevyengine/bevy/pull/5550

alice-i-cecile commented:
@dgriffin91 shall I close this out in favor of #5550?

DGriffin91 commented:
@alice-i-cecile Yep, I think that makes sense.
